grafana/grafana
# OpenTSDB: Convert the OpenTSDB Query Editor from Angular to React

This PR is to convert the OpenTSDB query editor to React. Fixes https://github.com/grafana/grafana/issues/50639

The [plan](https://docs.google.com/document/d/1HQ5xb8VZ0EKMgLtkacem1x5tJWjrD2Zs5lk81pz38Eo/edit#) is to change as little as possible but there may be small visual changes and component substitutions.

Testing

Compare this branch with the main branch. You can run both locally or you can run one docker version and this branch locally. In the custom.ini file you can change the http_port to 3001and the login_cookie_name so that you can switch between versions.

Run `make devenv sources=opentsdb` to have two provisioned opentsdb data sources with which to test. 

Manual test scenarios 
1. Run a query in Explore/Dashboards
- Select a metric, `cpu`
- Downsample
- The “fill” component is available for v2.3
- Add, edit, delete a filter for v2.3
- Add, edit, delete tags for v2.2 or less
2. Loading empty query in Explore/Dashboards
3. Loading existing query in Explore/Dashboards
4. Migrate a query from main to this branch. Create a query as shown above
- Switch branches and check that the query populates the query editor
5. Adding, removing, editing metric names (with autocomplete)
6. Check that the Select component used in the React version behaves similarly to the autocomplete in Angular
7. Adding, removing, editing tags/filters
8. Create variables and use them in metric names and function parameters
9. Error handling 
- Tags cannot be used with filters
- Two tags keys cannot be used

Notable changes
1. The metric input is changed to a select input to support autocomplete.
2. The suggested metrics are loaded once and filtered as opposed to before where they were loaded on every keystroke.
    -remove the autocomplete input
    -use a searchable select
4. The OpenTsdbQuery type has been updated to handle things that were previously defined in the target.
https://github.com/grafana/grafana/blob/7afd6bfa397e98b244dbcd857cd5ffe9b6102658/public/app/plugins/datasource/opentsdb/types.ts#L3-L20
5. Remove all angular code I come across, eg https://github.com/grafana/grafana/blob/c0c4409abc85276d49b81e8aecbfc7467a2c6362/public/app/plugins/datasource/opentsdb/datasource.ts#L540-L551
6. `suggestTagValues()` is called once now and filtered in a searchable select instead of being loaded every keystroke in the angular autocomplete (same as suggest metrics)
7. Small change to `suggestTagKeys()` in pass in query instead of metric to handle type issue.
